src/eval.rs

```diff
@@ -205,14 +205,21 @@ pub fn eval_main<'tcx>(tcx: TyCtxt<'tcx>, main_id: DefId, config: MiriConfig) ->
     // Perform the main execution.
     let res: InterpResult<'_, i64> = (|| {
         // Main loop.
-        while ecx.step()? {
+        loop {
+            match ecx.schedule()? {
+                SchedulingAction::ExecuteStep => {
+                    assert!(ecx.step()?, "a terminated thread was scheduled for execution");
+                }
+                SchedulingAction::ExecuteDtors => {
+                    ecx.schedule_next_tls_dtor_for_active_thread()?;
```

Comment:
```suggestion
                    // This will either enable the thread again (so we go back to `ExecuteStep`),
                    // or determine that this thread is done for good.
                    ecx.schedule_next_tls_dtor_for_active_thread()?;
```